src/node-http2-core.cc

```diff
@@ -0,0 +1,279 @@
+  if (chunks_cb == nullptr) {
+    chunks_cb = pending_data_chunks_free_list.pop();
+    chunks_cb->handle = stream_handle;
+    stream_handle->current_data_chunks_cb_ = chunks_cb;
+  }
+  nghttp2_data_chunk_t* chunk = data_chunk_free_list.pop();
+  chunk->buf = uv_buf_init(Malloc(len), len);
+  memcpy(chunk->buf.base, data, len);
+  if (chunks_cb->tail == nullptr) {
+    chunks_cb->head = chunk;
+    chunks_cb->tail = chunk;
+  } else {
+    chunks_cb->tail->next = chunk;
+    chunks_cb->tail = chunk;
+  }
+  return 0;
+}
+
+ssize_t Nghttp2Session::OnSelectPadding(nghttp2_session* session,
+                                        const nghttp2_frame* frame,
+                                        size_t maxPayloadLen,
+                                        void* user_data) {
+  Nghttp2Session* handle = static_cast<Nghttp2Session*>(user_data);
+  assert(handle->HasGetPaddingCallback());
+  return handle->GetPadding(frame->hd.length, maxPayloadLen);
+}
+
+ssize_t Nghttp2Session::OnStreamRead(nghttp2_session* session,
+                                     int32_t stream_id,
+                                     uint8_t* buf,
+                                     size_t length,
+                                     uint32_t* flags,
+                                     nghttp2_data_source* source,
+                                     void* user_data) {
+  Nghttp2Session* handle = static_cast<Nghttp2Session*>(user_data);
+  std::shared_ptr<Nghttp2Stream> stream_handle;
+  stream_handle = handle->FindStream(stream_id);
+  assert(stream_handle);
+
+  size_t remaining = length;
+  size_t offset = 0;
+
+  while (stream_handle->queue_head_ != nullptr) {
+    nghttp2_stream_write_queue* head = stream_handle->queue_head_;
+    for (unsigned int n = stream_handle->queue_head_index_;
+         n < head->nbufs; n++) {
+      if (head->bufs[n].len > 0) {
+        size_t len = head->bufs[n].len - stream_handle->queue_head_offset_;
+        len = len < remaining ? len : remaining;
+        memcpy(buf + offset,
+               head->bufs[n].base + stream_handle->queue_head_offset_,
+               len);
+        offset += len;
+        remaining -= len;
+        if (len < head->bufs[n].len) {
+          stream_handle->queue_head_offset_ += len;
+        } else {
+          stream_handle->queue_head_index_++;
+          stream_handle->queue_head_offset_ = 0;
+        }
+      } else {
+        goto end;
+      }
+    }
+    stream_handle->queue_head_offset_ = 0;
+    stream_handle->queue_head_index_ = 0;
+    stream_handle->queue_head_ = head->next;
+    head->cb(head->req, 0);
+    delete head;
+  }
+
+ end:
+  int writable = stream_handle->queue_head_ != nullptr ||
+                 stream_handle->IsWritable();
+  if (offset == 0 && writable && stream_handle->queue_head_ == nullptr) {
+    /* TODO(addaleax): ask @jasnell what the correct semantics are...
+       this is dead code right now */
+    return NGHTTP2_ERR_DEFERRED;
```

Comment:
@addaleax NGHTTP2_ERR_DEFERRED needs to be returned when the stream writable side has not been closed but there is no data currently available to send. What this return code does is tell nghttp2 to take this stream temporarily out of the send queue until data is available (which will happen on the next write).